pkg/cgroups/cgroupblkio.go

```diff
@@ -0,0 +1,168 @@
+	Major  int64
+	Minor  int64
+	Weight int64
+}
+
+// OciRateDeviceParameters contains values for
+// - blkio.throttle.read_bps_device
+// - blkio.throttle.write_bps_device
+// - blkio.throttle.read_iops_device
+// - blkio.throttle.write_iops_device
+type OciRateDeviceParameters struct {
+	Major int64
+	Minor int64
+	Rate  int64
+}
+
+// NewOciBlockIOParameters creates new OciBlockIOParameters instance.
+func NewOciBlockIOParameters() OciBlockIOParameters {
+	return OciBlockIOParameters{
+		Weight: -1,
+	}
+}
+
+// NewOciWeightDeviceParametes creates new OciWeightDeviceParameters instance.
+func NewOciWeightDeviceParametes() OciWeightDeviceParameters {
+	return OciWeightDeviceParameters{
+		Major:  -1,
+		Minor:  -1,
+		Weight: -1,
+	}
+}
+
+// NewOciRateDeviceParameters creates new OciRateDeviceParameters instance.
+func NewOciRateDeviceParameters() OciRateDeviceParameters {
+	return OciRateDeviceParameters{
+		Major: -1,
+		Minor: -1,
+		Rate:  -1,
+	}
+}
+
+// GetBlkioDir returns the cgroups blkio controller directory.
+func GetBlkioDir() string {
+	return blkioCgroupDir
+}
+
+// SetBlkioParameters writes OCI BlockIO parameters to files in cgroups blkio contoller directory.
+func SetBlkioParameters(cgroupsDir string, blockIO OciBlockIOParameters) error {
+	log.Debug("configuring cgroups blkio controller in directory %#v with parameters %+v", cgroupsDir, blockIO)
+	var errors *multierror.Error
+	if blockIO.Weight >= 0 {
+		errors = multierror.Append(errors, writeToFileInDir(cgroupsDir, blkioWeightFiles, strconv.FormatInt(blockIO.Weight, 10)))
+	}
+	for _, weightDevice := range blockIO.WeightDevice {
+		errors = multierror.Append(errors, writeDevValueToFileInDir(cgroupsDir, blkioWeightDeviceFiles, weightDevice.Major, weightDevice.Minor, weightDevice.Weight))
+	}
+	for _, rateDevice := range blockIO.ThrottleReadBpsDevice {
+		errors = multierror.Append(errors, writeDevValueToFileInDir(cgroupsDir, blkioThrottleReadBpsFiles, rateDevice.Major, rateDevice.Minor, rateDevice.Rate))
+	}
+	for _, rateDevice := range blockIO.ThrottleWriteBpsDevice {
+		errors = multierror.Append(errors, writeDevValueToFileInDir(cgroupsDir, blkioThrottleWriteBpsFiles, rateDevice.Major, rateDevice.Minor, rateDevice.Rate))
+	}
+	for _, rateDevice := range blockIO.ThrottleReadIOPSDevice {
+		errors = multierror.Append(errors, writeDevValueToFileInDir(cgroupsDir, blkioThrottleReadIOPSFiles, rateDevice.Major, rateDevice.Minor, rateDevice.Rate))
+	}
+	for _, rateDevice := range blockIO.ThrottleWriteIOPSDevice {
+		errors = multierror.Append(errors, writeDevValueToFileInDir(cgroupsDir, blkioThrottleWriteIOPSFiles, rateDevice.Major, rateDevice.Minor, rateDevice.Rate))
+	}
+	return errors.ErrorOrNil()
+}
+
+// writeDevValueToFileInDir writes MAJOR:MINOR VALUE to the first existing file under baseDir
+func writeDevValueToFileInDir(baseDir string, filenames []string, major, minor, value int64) error {
+	content := fmt.Sprintf("%d:%d %d", major, minor, value)
```

Comment:
why not to merge that with next function? in practice it is just one line with `fmt.Sprintf()` difference.

Reply:
When you reviewed the code, this and the unit tests for `cgroupsblkio.go` were not in place yet:

```go
type platformInterface interface {
    writeToFile(filename string, content string) error
}
```

The goal here is to keep the function that has to interact with the platform as simple as possible. This enables unit testing as much as possible, including this string formatting, just by mocking only the simple platform functions.

Currently `cgroupsblkio_test.go` verifies that correct string has been written to correct file.